### ISSUE TITLE
Tensorflow requirement issue

### ISSUE THREAD

**Issue by candalfigomoro** [SeldonIO/alibi-detect#387]
When I import alibi_detect, I get the following error:

`AttributeError: module 'tensorflow_core.keras.activations' has no attribute 'swish'`

which is related to this issue: https://github.com/huggingface/transformers/issues/7333

They say to update to TensorFlow 2.2, ~~but tf 2.2 requires at least Python 3.7.~~

~~So alibi-detect is basically not compatible with Python 3.6.~~

**Comment by candalfigomoro** [comment]
tf 2.2 should support Python 3.6, maybe tf 2.1 is enforced on my system by another dependency. I'll close the issue at the moment (more investigation is needed).

**Comment by ascillitoe** [comment]
Hi @candalfigomoro, thanks for clarifying! Our nightly CI is currently passing for Python 3.6 (https://github.com/SeldonIO/alibi-detect/runs/4258588653) so I agree it looks like it might be a specific tensorflow version issue on your end. If it's any help, our Python 3.6 CI installs `transformers==4.12.5` and `tensorflow==2.6.2`. 

Also just a friendly FYI, we will actually be dropping Python 3.6 very soon anyway as it is EOL on Dec 23rd. Please do keep us updated on this issue though, as it'll be good to know if there are any issues up until the time we drop it!

**Comment by candalfigomoro** [comment]
@ascillitoe 
I investigated the issue a little bit.

alibi-detect requires (https://github.com/SeldonIO/alibi-detect/blob/master/setup.py#L39):
```
        "tensorflow>=2.0.0, !=2.6.0, !=2.6.1, <2.7.0", # https://github.com/SeldonIO/alibi-detect/issues/375
        "tensorflow_probability>=0.8.0, <0.13.0",
        "transformers>=4.0.0, <5.0.0",
```

But there are some issues:

- Because of the issue linked above (https://github.com/huggingface/transformers/issues/7333), `transformers >3.1.0` requires `tensorflow >=2.2`. The required version in alibi-detect is `transformers>=4.0.0, <5.0.0`, so tensorflow versions <2.2 don't work here.

-  `transformers >=4.2.0` requires `tensorflow >=2.3`, but this requirement is correctly managed by the transformers package (https://github.com/huggingface/transformers/blob/v4.2.0/setup.py#L131), so this is not an issue.

- alibi-detect requires `tensorflow_probability>=0.8.0, <0.13.0`, but versions `>=0.12` (and `<0.13.0`) include a runtime check for `tensorflow >=2.4` (see https://github.com/tensorflow/probability/blob/v0.12.0/tensorflow_probability/python/__init__.py#L58). So alibi-detect is broken if you use tensorflow_probability 0.12 with tensorflow < 2.4.

Basically, to avoid issues with dependencies, the tensorflow requirement for alibi-detect could be raised to `tensorflow >=2.4`.

**Comment by ascillitoe** [comment]
@candalfigomoro thank you very much for looking into this in so much detail, this is really helpful.

With respect to these issues, I think two of them are resolved OK by pip, whereas you seem to have uncovered a definite issue with the third!

1. pip should enforce `tensorflow >=2.2`due to `transformers>=4.0.0, <5.0.0` (since `transformers` specs `tensorflow`/`tensorflow-cpu` in `setup.py`). Although we could be more explict here and bump to `tensorflow>=2.2` in our `setup.py`.
2. As you say, similar story here. 
3. This is a problem. It arises because `tensorflow-probablity` does not actually spec `tensorflow` in `setup.py`. Related to https://github.com/tensorflow/probability/issues/1468. Hopefully there might be a solution to this on the `tensorflow-probability` side at some point, but right now I suspect you might be right that setting `tensorflow>=2.4` on our end is the safest solution.

Grateful for your thoughts on the above, especially if you disagree with anything!

**Comment by ascillitoe** [comment]
P.s. bumping to `tensorflow>=2.4.0` requires more discussion. It would restrict users to what is quite a new version so we need to think more about this.

**Comment by candalfigomoro** [comment]
> 1. pip should enforce `tensorflow >=2.2`due to `transformers>=4.0.0, <5.0.0` (since `transformers` specs `tensorflow`/`tensorflow-cpu` in `setup.py`). Although we could be more explict here and bump to `tensorflow>=2.2` in our `setup.py`.

@ascillitoe 
About this point, tensorflow >=2.3 is enforced for transformers >=4.2.0. For versions 4.0.x and 4.1.x (which are allowed by alibi-detect), it requires tensorflow >= 2.0 (see https://github.com/huggingface/transformers/blob/v4.1.1/setup.py#L130). The problem is that tensorflow 2.0.x and 2.1.x are broken because of the linked issue (https://github.com/huggingface/transformers/issues/7333). This is a transformers issue, but it indirectly affects alibi-detect. Basically, alibi-detect is broken when you use tensorflow 2.0 or 2.1 with transformers 4.0 or 4.1.

**Comment by ascillitoe** [comment]
@candalfigomoro understood! Thank you for clearing that up. So we do actually have two distinct issues here. It seems like regardless of the solution we choose for `tensorflow-probability we might want to explicitly bump to `tensorflow>=2.2`.

**Comment by candalfigomoro** [comment]
@ascillitoe 
Yes.

As a side note, I had the need to use alibi-detect with tensorflow 2.1 (because of other constraints). Since the latest transformers version that works with tf 2.1 is transformers 3.1.0, I "downgraded" the alibi-detect requirement for transformers (>=4.0.0) to also accept transformers version 3.1.0. I don't know if everything works correctly, but at least I'm able to load alibi-detect and use the TabularDrift class (which doesn't rely on tensorflow).

**Comment by ascillitoe** [comment]
#404 bumps the tensorflow lower bound to 2.2. If/when we modularise `transformers` in the future (https://github.com/SeldonIO/alibi-detect/issues/388), we can relax the `tensorflow` bounds.

I'll leave this issue open as there is still the other issue with `tensorflow-probability`.

**Comment by ascillitoe** [comment]
@candalfigomoro after running some tests with the `tensorflow_probablity` and `tensorflow` versions you mentioned (e.g. 0.12 and 2.0.0), it looks to me like `tensorflow_probability` raises quite a descriptive `ImportError`, so I'm not sure we would add much by raising our own errors on top of this. We think it would be undesirable to move to `tensorflow>=2.4` at this point, as it is still relatively new and lots of users might still be on older `tensorflow`/`tensorflow_probability` versions. 

Hopefully in the future the `tensorflow` version will be specified by `tensorflow_probability` and this issue will be circumvented entirely 🤞🏻 

I'll close this issue now, but please do feel free to reopen if you think it still isn't resolved!
